specifyweb/frontend/js_src/lib/components/DataModel/uniquenessRules.ts

```diff
@@ -0,0 +1,184 @@
+import React from 'react';
+
+import { useCachedState } from '../../hooks/useCachedState';
+import { formsText } from '../../localization/forms';
+import { ajax } from '../../utils/ajax';
+import { getCache, setCache } from '../../utils/cache';
+import type { GetOrSet, RA } from '../../utils/types';
+import { formatConjunction } from '../Atoms/Internationalization';
+import { load } from '../InitialContext';
+import type { SerializedResource } from './helperTypes';
+import type { LiteralField, Relationship } from './specifyField';
+import type { SpLocaleContainerItem, Tables } from './types';
+
+export type UniquenessRules = {
+  readonly [TABLE in keyof Tables]?: RA<{
+    readonly id: number | null;
+    readonly fields: RA<SerializedResource<SpLocaleContainerItem>>;
+    readonly scope: SerializedResource<SpLocaleContainerItem> | null;
+    readonly isDatabaseConstraint: boolean;
+
+    // This property is assigned on the frontend and is not saved to the backend
+    readonly uniqueId?: number;
+  }>;
+};
+
+export type UniquenessRule = Exclude<
+  UniquenessRules[keyof Tables],
+  undefined
+>[number];
+
+const setInitialRules = async (): Promise<UniquenessRules> =>
+  import('./schema')
+    .then(async ({ fetchContext }) => fetchContext)
+    .then(async (schema) =>
+      load<UniquenessRules>(
+        `/businessrules/uniqueness_rules/${schema.domainLevelIds.discipline}/`,
+        'application/json'
+      )
+    )
+    .then((data) => {
+      setCache(
+        'businessRules',
+        'uniqueRules',
+        data as Readonly<Record<string, string>>
+      );
+      return data;
+    });
+
+export const fetchContext = setInitialRules();
+
+export function getUniquenessRules(): UniquenessRules | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  model: TABLE_NAME
+): UniquenessRules[TABLE_NAME] | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  model?: TABLE_NAME
+): UniquenessRules | UniquenessRules[TABLE_NAME] {
+  const uniquenessRules = getCache('businessRules', 'uniqueRules');
```

Comment:
Yeah I don't think it is necessary to cache uniqueness rules. Rather, can they be stored in an object like preferences? 

A possible complication can be if user opens multiple tabs and modifies uniqueness rules in one tab, but doesn't refresh the second tab. However, since the uniqueness rule is also saved on the backend, it won't cause a data integrity issue in the second tab. So, in short, not necessary to cache them.

Reply:
Removed cache in https://github.com/specify/specify7/pull/4215/commits/4444e81b6cc70ba33a3cf77b223859a0ee7dd6ae